src/RADCodeGen.hs

```diff
@@ -0,0 +1,557 @@
+module RADCodeGen where
+  import Grammar
+  import LALR
+  import GenUtils (mapDollarDollar)
+  import RADTools (XGrammar(..), showItem, showProd, lhs, showRecognitionPoint, recognitionPoints, rhsAfterDot)
+  import RADStateGen
+  import Control.Monad
+  import Data.List
+  import Data.Maybe
+  import Data.Text (pack, unpack, replace)
+  import GHC.Arr ((!), indices)
+  
+
+  data ParserType = Normal | Monad | MonadLexer deriving (Eq, Show)
+  
+  data GenOptions = GenOptions {
+    ptype :: ParserType,
+    
+    wrapperType :: String, -- e.g. "Parser"
+    errorTokenType :: String, -- e.g. "ErrorToken"
+    
+    showTypes :: Bool,
+    comments :: Bool,
+    
+    rank2Types :: Bool, -- when True, all functions (including goto functions) which use or enclose a higher-rank-function are annotated with an explicit type.
+    forallMatch :: String, -- the text which determines which types count as rank-2-types.
+    
+    rulesTupleBased :: Bool, -- actually, this doesn't produce good nor fast code. maybe drop this?
+
+    optimize :: Bool, -- inline all rule functions and eta-expand all applications of local "g" functions
+
+    header :: String,
+    footer :: String
+  } deriving Show
+  
+  mlex opts = ptype opts == MonadLexer
+  
+  raw = flip (.) _raw
+  
+  dotleft (Lr0 rule dot) = Lr0 rule (dot-1)
+  
+  hasNT state = elem (raw radType state) [Type1, Type2]
+  
+  replaceDollar a char = maybe a ($ char) (mapDollarDollar a)
+  
+  -------------------- GENCODE --------------------
+  -- Generate the full code
+  genCode :: GenOptions -> XGrammar -> [RADState] -> ActionTable -> GotoTable -> IO String
+  genCode opts x states action goto = do
+    return $ newlines 3 [languageFeatures, header', entryPoints', definitions', rules', parseNTs', parseTerminals', states', actions', footer'] where
+      languageFeatures
+        | rank2Types opts = newline $ map extension ["RankNTypes", "ScopedTypeVariables"]
```

Comment:
I think that the user should activate `-XRankNTypes` if needed. Although I'm not so sure about `ScopedTypeVariables`...

Maybe you should grep for a line starting with `{-#` and cointaining `RankNTypes` to insert `ScopedTypeVariables` and to activate the annotation of type signatures in general.